File: typescript/src/02-factory/after/checkout-service.ts
import type { PaymentResult } from '../../shared/models.js';
import type { PaymentStrategyFactory } from './payment-strategy-factory.js';

// Checkout orchestrates the workflow. It does NOT instantiate objects itself.
export class CheckoutService {
  constructor(private readonly factory: PaymentStrategyFactory) {}

  checkout(method: string, amount: number): PaymentResult {
    return this.factory.create(method).pay(amount); // pick + run
  }
}

File: csharp/DesiArchitect.DesignPatterns/02_Factory/After/CheckoutService.cs
using DesiArchitect.DesignPatterns.Shared;

namespace DesiArchitect.DesignPatterns.Factory.After;

public class CheckoutService
{
    private readonly PaymentStrategyFactory _factory;

    public CheckoutService(PaymentStrategyFactory factory) => _factory = factory;

    public PaymentResult Checkout(string method, decimal amount)
        => _factory.Create(method).Pay(amount);
}
